core/translate.c

```diff
@@ -1972,10 +1978,13 @@ stress_test_recreate_state(dcontext_t *dcontext, fragment_t *f, instrlist_t *ili
             }
 
             if (spill_ibreg_outstanding_offs != UINT_MAX) {
-                mc.IF_X86_ELSE(xcx, r2) =
+                /* FIXME i#3544: RFC: Perhaps we need a union field for IBL register in
```

Comment:
Not necessarily. Consider how LR is defined [in `mxatx_api.h`](https://github.com/DynamoRIO/dynamorio/blob/master/core/lib/mcxtx_api.h#L87):
```c
    union {
        reg_t r30; /**< The r30 register. \note For 64-bit DR builds only. */
        reg_t lr;  /**< The link register. */
    }; /**< The anonymous union of alternative names for r30/lr register. */
```
What I suggest is to have a similar union around `xcx` and `r2`, so for X86:
```c
    union {
        reg_t xcx; /**< The platform-independent name for full rcx/ecx register. */
        reg_t IF_X64_ELSE(rcx, ecx); /**< The platform-dependent name for
                                          rcx/ecx register. */
        reg_t ibl; /**< The platform-independent name for the Indirect Branch Lookup target register */
    }; /**< The anonymous union of alternative names for rcx/ecx register. */
```
For ARM:
```c
    union {
        reg_t r2;   /**< The r2 register. */
        reg_t ibl; /**< The platform-independent name for the Indirect Branch Lookup target register */
    }; /**< The anonymous union of alternative names for r2 register. */
```
And for RISC-V:
```c
    union {
        reg_t x12; /**< The x12 register. */
        reg_t a2;  /**< The 3rd argument register. */
        reg_t ibl; /**< The platform-independent name for the Indirect Branch Lookup target register */
    };  /**< The anonymous union of alternative names for the x12/a2 register. */
```

Using the above line 1975 in `translate.c` would look like this:
```c
mc.ibl = (reg_t)d_r_get_tls(spill_ibreg_outstanding_offs) + 1;
```
I'll adapt to this change in the next version to show exactly what I mean.

Reply:
I think core/lib/mcxtx_api.h is expected to define the machine's context only, independent of how DR is using it. By adding the definitions you suggested, we're including details about how DR is using the available regs to store the IBL target.

IMHO defining the `IBL(mc)` macro (in source outside the header) is a better way to accomplish code simplicity.

Reply:
Since this came up in another comment thread, I think it'll be good to get @derekbruening 's opinion. I believe it'll be better to keep structs like spill_state_t and mcxtx independent of DR's implementation details.

Reply:
I agree that macros will leave the structs more readable at the first glance and also containing only HW-dependent fields.
It'll be then easier to move the arch-specific functional macros to `core/arch/<arch>/some_header.h`.